### PR TITLE
[Property Wrappers] Fix computation of lvalue-ness in buildStorageReference

### PR DESCRIPTION
**Pull request by hborla**
This picks up https://github.com/apple/swift/pull/29069 to resolve merge conflicts, fix build failures due to request evaluator changes, and fix one more case that crashed. Below is the original PR description. Thank you to @roop for all of the work here!

---

This is a follow-up to PR #27727 by @theblixguy to fix l-value-ness computation when building a reference to the storage of properties when synthesising accessors.

This PR fixes the following crashes:
1. Crash when using a property wrapper with a mutating get and a non-mutating set (SR-11637)
2. Crash when using a property wrapper with a mutating get, composed with any outer wrapper

Resolves SR-11637.
Resolves rdar://problem/60606399, rdar://problem/56467140

[SR-11637]: https://bugs.swift.org/browse/SR-11637

### Implementation

The `buildStorageReference` function in TypeCheckStorage.cpp builds references to the storage of a property. The references it builds are used in property accessors synthesised automatically by the compiler.

`buildStorageReference` needs to create expressions of the following forms:

- Without property wrappers:
    - Stored: self.member
- With property wrappers:
    - Wrapped: self._member.wrappedValue
    - Wrapped with composition: self._member.wrappedValue.wrappedValue….wrappedValue
    - Projected: self._member.projectedValue
    - Enclosed instance: Wrapper[_enclosedInstance: self, …]

With this PR, the l-valueness of different parts of these expressions are determined thus:

1. Whether the last component in the expression is an l-value or not is based on the `isLValue` argument to buildStorageReference
2. Whether the self component in the expression (except for the use in the `_enclosedInstance:` parameter above) is an l-value or not should be determined based on whether the accessor being synthesised is a mutating accessor or not. The `isSelfReference` local variable is used to store this decision.
3. All the other (middle) components would be instances of property wrappers. Whether such a component is an l-value or not is determined based on whether the access to the property wrapper instance’s `wrappedValue` (or `projectedValue`) requires l-valueness or not. The `underlyingVars` local variable now holds a second field for each entry to store this decision.

In order to determine (2) and (3) above, buildStorageReference now splits its `isLValue` argument into two arguments: `isUsedForGetAccess` and `isUsedForSetAccess` (commit a6e4b1a).

To determine (3) above, a new `PropertyWrapperLValueness` struct is computed, which holds information on which property wrappers in the chain need to be l-values to be able to get / set the final wrapped value (commit 66d186e).

### Testing

This PR just tests that there’s no crash while emitting SIL when using non-default mutatingness with wrappedValue, projectedVaue and composition.

I’m not sure if it’s worthwhile to check for the correct l-valueness for each component of the expression at AST stage by adding tests to test/decl/var/property_wrappers_synthesis.swift. Please let me know if you think that would be useful.

I also haven’t added tests for the `subscript(_enclosedInstance:…)` form because:
  - The mutatingness can affect the built reference only when there’s composition
  - I assume this form isn’t meant to be used with composition by design because there’s no good way to define the `wrappedKeyPath`’s type when used with composition of wrappers

### CODE REVIEW

**Comment by hborla** [comment]
@swift-ci please smoke test

**Comment by hborla** [comment]
@swift-ci please test source compatibility

File: lib/Sema/TypeCheckStorage.cpp [hunk 1]
```diff
@@ -341,8 +341,16 @@ IsSetterMutatingRequest::evaluate(Evaluator &evaluator,
...
-      return mut->Setter == PropertyWrapperMutability::Mutating
-        && result;
+      bool isMutating = mut->Setter == PropertyWrapperMutability::Mutating;
+      if (auto *accessor = var->getParsedAccessor(AccessorKind::DidSet)) {
+        // If there's a didSet, we call the getter for the 'oldValue', and so
+        // should consider the getter's mutatingness as well
+        if (!accessor->isSimpleDidSet()) {
+          isMutating |= (mut->Getter == PropertyWrapperMutability::Mutating);
+        }
+        isMutating |= (accessor->getAttrs().getAttribute<MutatingAttr>() != nullptr);
```
Comment (slavapestov) [inline lib/Sema/TypeCheckStorage.cpp:351]: You can use hasAttribute() instead

File: lib/Sema/TypeCheckStorage.cpp [hunk 2]
```diff
@@ -587,20 +595,34 @@ getEnclosingSelfPropertyWrapperAccess(VarDecl *property, bool forProjected) {
...
+                                   bool isUsedForGetAccess,
+                                   bool isUsedForSetAccess,
                                    ASTContext &ctx) {
+  // Whether the last component of the expression should be an l-value
+  bool isLValue = isUsedForSetAccess;
   // Local function to "finish" the expression, creating a member reference
   // to the given sequence of underlying variables.
   Optional<EnclosingSelfPropertyWrapperAccess> enclosingSelfAccess;
-  llvm::TinyPtrVector<VarDecl *> underlyingVars;
+  llvm::SmallVector<std::pair<VarDecl *, bool>, 1> underlyingVars;
```
Comment (slavapestov) [inline lib/Sema/TypeCheckStorage.cpp:621]: A comment explaining what the bool is for would be nice

File: lib/Sema/TypeCheckStorage.cpp [hunk 3]
```diff
@@ -1686,16 +1745,19 @@ synthesizeCoroutineAccessorBody(AccessorDecl *accessor, ASTContext &ctx) {
...
+  if (isModify &&
       (storageReadWriteImpl == ReadWriteImplKind::StoredWithSimpleDidSet ||
        storageReadWriteImpl == ReadWriteImplKind::InheritedWithSimpleDidSet)) {
     return synthesizeModifyCoroutineBodyWithSimpleDidSet(accessor, ctx);
   }
 
   // Build a reference to the storage.
-  Expr *ref = buildStorageReference(accessor, storage, target, isLValue, ctx);
+  Expr *ref = buildStorageReference(accessor, storage, target,
+                /*isUsedForGetAccess=*/true,
```
Comment (slavapestov) [inline lib/Sema/TypeCheckStorage.cpp:1758]: Nitpick, but this should line up with the (

**Review by slavapestov — COMMENTED** [review]
The PR description is great. Perhaps part of it should become a doc comment for buildStorageReference()?

**Comment by roop** [comment]
Here's an explanation of the algorithm used in `PropertyWrapperLValuenessRequest::evaluate`:

-----

The goal is to figure out whether accessing the `wrappedValue` of the
property wrapper instance would require that property wrapper instance
to be an l-value or not.

If we have Swift code like `@A @B @C var member: Int`, we'll be building
an expression like
`self._member.wrappedValue.wrappedValue.wrappedValue`.

~~~
@A @B @C var member: Int

self    _member     wrappedValue    wrappedValue    wrappedValue
        (of type A) (of type B)     (of type C)     (of type Int)

       |                                        |
       +----------------------------------------+
       Property wrapper instances
~~~

For each of these property wrapper instances, we're trying to figure out
whether accessing its `wrappedValue` property (for either get or set)
requires the property wrapper instance to be an l-value or not. This is
what `PropertyWrapperLValueness` describes.

Calling the getter (or setter) on the nth property wrapper in the chain
is done as follows:
 1. call the getter on the (n-1)th property wrapper instance to get the
    nth property wrapper instance
 2. call the getter (or setter) on the nth property wrapper instance
 3. if (2) is a mutating access, call the setter on the (n-1)th property
    wrapper instance to write back the mutated value

So, a straightforward algorithm to compute l-valueness would look like this:

~~~
func getLValueness() -> (forGetAccess: LValueness, forSetAccess: LValueness) {
  lvaluenessForGet: [Bool] = []
  lvaluenessForSet: [Bool] = []
  lvaluenessForGetOrSet: [Bool] = []

  for n in 0 ..< numWrappers {

    currentGetSequence = isGetterMutating(at: n) ?
          lvaluenessForGetOrSet  // Both get and set will be used on the (n-1)th instance
                    + [true]     // Get access on nth instance requires lvalue
        : lvaluenessForGet       // Only get will be used on the (n-1)th instance
                    + [false]    // Get access on nth instance doesn't require lvalue

    currentSetSequence = isSetterMutating(at: n) ?
          lvaluenessForGetOrSet  // Both get and set will be used on the (n-1)th instance
                    + [true]     // Set access on nth instance requires lvalue
        : lvaluenessForGet       // Only get will be used on the (n-1)th instance
                    + [false]    // Set access on nth instance doesn't require lvalue

    currentGetOrSetSequence = (isGetterMutating(at: n) || isSetterMutating(at: i)) ?
          lvaluenessForGetOrSet  // Both get and set will be used on the (n-1)th instance
                    + [true]     // Get or set access on nth instance requires lvalue
        : lvaluenessForGet       // Only get will be used on the (n-1)th instance
                    + [false]    // Neither get nor set on nth instance requires lvalue

    lvaluenessForGet = currentGetSequence
    lvaluenessForSet = currentSetSequence
    lvaluenessForGetOrSet = currentGetOrSetSequence

  }

  return (lvaluenessForGet, lvaluenessForSet)
}
~~~

The above algorithm might however be bad on performance because:

  - we're creating, copying and destroying vectors of booleans in every iteration
  - `currentSetSequence` is computed in every iteration but only the
    last computed value is ever used

So here's an optimized version:

~~~
func getLValueness() -> (forGetAccess: LValueness, forSetAccess: LValueness) {
  previousLvaluenessForGet: [Bool] = []
  previousLvaluenessForGetOrSet: [Bool] = []
  isCurrentGetterMutating: Bool = isGetterMutating(at: 0)
  isCurrentSetterMutating: Bool = isSetterMutating(at: 0)

  for i in 1 ..< numWrappers {

    n = i - 1 // n is used only in the comments
    if isCurrentGetterMutating {
      // if getter is mutating at nth instance, then get access on the
      // nth instance requires both get and set on the (n-1)th instance
      previousLvaluenessForGet = previousLvaluenessForGetOrSet
    }
    if (!isCurrentGetterMutating && !isCurrentSetterMutating) {
      // if both getter and setter are nonmutating at nth instance,
      // then both get and set access on the nth instance require only
      // get access on the (n-1)th instance
      previousLvaluenessForGetOrSet = previousLvaluenessForGet
    }

    // In case
    // isCurrentGetterMutating = false && isCurrentSetterMutating == true
    // then: 
    //   - get access on nth instance requires only get on the 
    //     (n-1)th instance
    //   - get + set access on the nth instance requires both get and
    //     set on the (n-1)th instance
    //
    // So previousLvaluenessForGet and previousLvaluenessForGetOrSet can
    // remain unchanged.

    // Append the lvalueness for accessing the nth instance
    previousLvaluenessForGet.append(isCurrentGetterMutating)
    previousLvaluenessForGetOrSet.append(isCurrentGetterMutating || isCurrentSetterMutating)

    // Refresh for next iteration
    isCurrentGetterMutating = isGetterMutating(at: i)
    isCurrentSetterMutating = isSetterMutating(at: i)

  }

  // For the last lap, we compute lvalueness for get and set

  lvaluenessForGet = isCurrentGetterMutating ?
          previousLvaluenessForGetOrSet + [true]
        : previousLvaluenessForGet + [false]

  lvaluenessForSet = isCurrentSetterMutating ?
          previousLvaluenessForGetOrSet + [true]
        : previousLvaluenessForGet + [false]

  return (lvaluenessForGet, lvaluenessForSet)
}
~~~

The PR contains the optimized version.

-----

**Comment by hborla** [comment]
@roop Thank you! I understand. What was confusing me previously was why we needed to potentially change the mutability of every previous access at each iteration:
```cpp
      if (mutability.Getter == PropertyWrapperMutability::Mutating) {
        prevWrappersMutabilityForGet = prevWrappersMutabilityForGetAndSet;
      }
      if (mutability.Getter != PropertyWrapperMutability::Mutating &&
          mutability.Setter != PropertyWrapperMutability::Mutating) {
        prevWrappersMutabilityForGetAndSet = prevWrappersMutabilityForGet;
      }
```
Since the mutability of the `n-1`th access depends on the mutability of the `n`th access, it seems like this could be simplified by building up the mutability for get and set sequences directly by iterating from the innermost wrapper to the outermost. Am I missing a reason why we need to iterate from outermost --> innermost?

**Comment by slavapestov** [comment]
I'm curious if the algorithm actually has any intrinstic dependency on property wrappers, or if it could be generalized to computing the access kind for an arbitrary "path" of lvalues.

I ask because Sema has several incomplete implementations of this in a few places that are just good enough to produce diagnostics, and LoadExpr/InOutExpr nodes in the AST; SILGen has a completely different method of doing this which is 100% accurate but operates on SILGen's structures. If we can centralize this in one place in the AST we can delete some code and get more accurate diagnostics.

Here's an example where the current code goes wrong:

```
class O {
    var string = "" {
        willSet {
            let oldCount = string.utf8.count
            let newCount = newValue.utf8.count

            print(oldCount, newCount)
        }
    }
}
```

We produce a bogus diagnostic "attempting to store to property 'string' within its own willSet, which is about to be overwritten by the new value" which clearly makes no sense. We're not storing into "string" at all, we're accessing a member.

Now I'm obviously not suggesting fixing this all in this PR, but it's just a future direction to consider.

**Comment by hborla** [comment]
@swift-ci please smoke test

**Comment by hborla** [comment]
@swift-ci please test source compatibility

**Comment by hborla** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckStorage.cpp [hunk 4]
```diff
@@ -341,8 +341,16 @@ IsSetterMutatingRequest::evaluate(Evaluator &evaluator,
   // or not based on the composition of the wrappers.
   if (auto var = dyn_cast<VarDecl>(storage)) {
     if (auto mut = var->getPropertyWrapperMutability()) {
-      return mut->Setter == PropertyWrapperMutability::Mutating
-        && result;
+      bool isMutating = mut->Setter == PropertyWrapperMutability::Mutating;
+      if (auto *accessor = var->getParsedAccessor(AccessorKind::DidSet)) {
+        // If there's a didSet, we call the getter for the 'oldValue', and so
+        // should consider the getter's mutatingness as well
+        if (!accessor->isSimpleDidSet()) {
```
Comment (DougGregor) [inline lib/Sema/TypeCheckStorage.cpp:348]: Is there a similar issue with `willSet`?

File: lib/Sema/TypeCheckStorage.cpp [hunk 5]
```diff
@@ -701,22 +737,42 @@ static Expr *buildStorageReference(AccessorDecl *accessor,
...
-      // Check for availability of wrappedValue.
-      if (accessor->getAccessorKind() == AccessorKind::Get ||
-          accessor->getAccessorKind() == AccessorKind::Read) {
-        if (wrappedValue->getAttrs().getUnavailable(ctx)) {
-          diagnoseExplicitUnavailability(
-              wrappedValue,
-              var->getAttachedPropertyWrappers()[i]->getRangeWithAt(),
-              var->getDeclContext(), nullptr);
+    if (firstWrapperIdx < lastWrapperIdx) {
+      if (auto lvalueness = getPropertyWrapperLValueness(var)) {
```
Comment (DougGregor) [inline lib/Sema/TypeCheckStorage.cpp:741]: `firstWrapperIdx < lastWrapperIdx` guarantees that `getPropertyWrapperLValueness` will return a value, so this `if` can be replaced with

```
auto lvalueness = *getPropertyWrapperLValueness(var);
```

right?

**Review by DougGregor — APPROVED** [review]
This looks fantastic!

File: lib/Sema/TypeCheckStorage.cpp (hunk 4 above)
Comment (hborla) [reply lib/Sema/TypeCheckStorage.cpp:348]: I don't think so - the getter is called for the argument to `didSet`, while `willSet` just takes the setter argument. However, there's probably a similar issue wrt the mutability of `willSet`

**Comment by hborla** [comment]
@swift-ci please smoke test

**Comment by hborla** [comment]
@swift-ci please smoke test Linux platform

**Comment by roop** [comment]
@hborla Your simplification of the algorithm is awesome. I don't quite understand how it works yet, but I can see it works very well.

**Comment by hborla** [comment]
@swift-ci please smoke test Linux platform

**Comment by roop** [comment]
@slavapestov I think the core idea of the algorithm is not dependent on property wrappers, and can work for an arbitrary sequence of `MemberRefExpr`-essions.

> I ask because Sema has several incomplete implementations of this in a few places that are just good enough to produce diagnostics, and LoadExpr/InOutExpr nodes in the AST;

When we’re accessing a chain of `MemberRefExpr`-essions like `a.b.c.d` and `a`, `b` and `c` are instances of structs, then we might need a writeback -- if c’s d-accessor is mutating, then c needs to be written back onto b, and so on. Is this what you're talking about here?

I think all these instances are conservatively assumed to be lvalues at present, which is what is causing the bogus warning (the compiler incorrectly thinks “string” is used as an lvalue, and warns that it shouldn’t be written to inside a willSet). Ideally, an algorithm similar to what’s in this PR should be used to decide whether these instances should be lvalues or not. Is that a correct assessment?

> SILGen has a completely different method of doing this which is 100% accurate but operates on SILGen's structures.

Can you please point me to this SILGen implementation?

**Pull request merged**